### DISCUSSION TITLE
`@apply` not being implemented upon deplyoment in Vercel

### DISCUSSION THREAD

**Discussion by lng-athens** [tailwindlabs/tailwindcss#14803] [Help]
I have recently started a project using the `npm create vite@latest my-project -- -- template react` as stated in the [Vite Guide](https://vite.dev/guide/). So far I have been doing well with the project in development running in localhost. However, when I now tried to deploy it in Vercel, some of the TailwindCSS styles are not being implemented, specifically the ones with the `@apply` styling of Tailwind. Here is an example:

`\src\assets\css\login.css`
```
@tailwind base;
@tailwind components;
@tailwind utilities;

#app-container {
  grid-template-columns: 1fr;
  grid-template-rows: 1fr;

  .content {
    @apply flex items-center justify-center;
    grid-column: auto;
    grid-row: auto;
    background-image: url('/images/backgrounds/techno-blue-hexagonal-wave.jpg');
    background-position: center;
    background-repeat: no-repeat;
    background-size: cover;

    .login-container {
        @apply justify-center w-full px-4;
        
        .col {
            @apply rounded-md py-4;
            background-color: rgba(0, 0, 0, 0.5);

            .logo {
                @apply mt-3 mb-4 mx-auto w-2/3;
            }

            .form-group {
                @apply px-4 mx-auto;

                .form-label {
                    @apply text-text font-mono text-sm;
                }
    
                .form-control {
                    @apply font-mono;
                }
            }

            .btn-submit {
                @apply text-text font-bold mx-auto mt-8 block px-10;
                background-color: #0d6efd;
            }
        }
    }
  }
}

@media (min-width: 1200px) {
    .login-container {
        .form-group {
            @apply px-0;
            width: 90%;
        }
    }
}
```

The style `@apply` is not being implemented but the rest of the manual styling is working fine. I also have the `base`, `components`, `utilities` from `@tailwind` imported in the `\src\index.css` of my project.

Here is the configs as I have followed them in the [TailwindCSS Guide](https://tailwindcss.com/docs/installation/using-postcss).
`\postcss.config.js`
```
export default {
  plugins: {
    'postcss-import': {},
    tailwindcss: {},
    autoprefixer: {},
  },
};
```

`\tailwind.config.js`
```
/** @type {import('tailwindcss').Config} */
export default {
  content: [
    './index.html',
    './src/**/*.{js,jsx,ts,tsx}',
    './src/pages/**/*.{js,jsx,ts,tsx}',
    './src/components/**/*.{js,jsx,ts,tsx}',
    './src/assets/css/**/*.css',
  ],
  theme: {
    colors: {
      primary: '#2a2a2a',
      'primary-rgb': '42, 42, 42',
      secondary: '#39ff14',
      'secondary-rgb': '57, 255, 20',
      tertiary: '#2ed610',
      'tertiary-rgb': '46, 214, 16',
      accent: '#ff0000',
      'accent-rgb': '255, 0, 0',
      background: '#1e1e1e',
      'background-rgb': '30, 30, 30',
      text: '#f5f5f5',
      'text-rgb': '245, 245, 245',
      danger: '#dc3545'
    },
    fontFamily: {
      serif: ['EB Garamond', 'serif'],
      sans: ['Neuton', 'sans-serif'],
      mono: ['Courier Prime', 'monospace'],
      cursive: ['Dancing Script', 'cursive'],
    },
    extend: {},
  },
  plugins: [],
};
```

`\package.json`
```
{
  "name": "project-name",
  "private": true,
  "version": "1.0.0",
  "description": "Something.",
  "type": "module",
  "scripts": {
    "dev": "vite",
    "build": "vite build",
    "lint": "eslint .",
    "preview": "vite preview"
  },
  "repository": {
    "type": "git",
    "url": "my-git-repo"
  },
  "dependencies": {
    "@fortawesome/fontawesome-svg-core": "^6.6.0",
    "@fortawesome/free-brands-svg-icons": "^6.6.0",
    "@fortawesome/free-regular-svg-icons": "^6.6.0",
    "@fortawesome/free-solid-svg-icons": "^6.6.0",
    "@fortawesome/react-fontawesome": "^0.2.2",
    "@lordicon/react": "^1.10.0",
    "aos": "^2.3.4",
    "axios": "^1.7.7",
    "js-cookie": "^3.0.5",
    "lottie-web": "^5.12.2",
    "react": "^18.3.1",
    "react-dom": "^18.3.1",
    "react-hook-form": "^7.53.1",
    "react-router-dom": "^6.26.2",
    "react-scroll": "^1.9.0",
    "react-svg": "^16.1.34",
    "sweetalert2": "^11.14.2"
  },
  "devDependencies": {
    "@eslint/js": "^9.11.1",
    "@types/node": "^22.7.5",
    "@types/react": "^18.3.10",
    "@types/react-dom": "^18.3.0",
    "@vitejs/plugin-react": "^4.3.2",
    "autoprefixer": "^10.4.20",
    "cssnano": "^7.0.6",
    "eslint": "^9.11.1",
    "eslint-plugin-react": "^7.37.0",
    "eslint-plugin-react-hooks": "^5.1.0-rc.0",
    "eslint-plugin-react-refresh": "^0.4.12",
    "globals": "^15.9.0",
    "postcss": "^8.4.47",
    "postcss-import": "^16.1.0",
    "tailwindcss": "^3.4.13",
    "vite": "^5.4.8"
  }
}
```

I don't know what's wrong or missing but when I deploy it in [Vercel](https://vercel.com/), the ones with the `@apply` is not working.

**Comment by aniruddhaadak80** [comment]
Hi @lng-athens,

It looks like your `@apply` styles aren't working on Vercel. Here are some quick steps to troubleshoot:

1. **Run Production Build**: Locally run `npm run build` and preview with `npm run preview` to check if the issue exists before deploying.

2. **Check CSS Imports**: Ensure your CSS files are correctly imported in `index.css`:
   ```css
   @import './assets/css/login.css';
   ```

3. **Verify Tailwind Config**: Ensure your `tailwind.config.js` correctly lists all paths in the `content` array.

4. **Install Dependencies**: Run `npm install` to ensure all dependencies are installed.

5. **Clear Vercel Cache**: In Vercel settings, clear any caches and redeploy.

6. **Inspect Errors**: Check the developer console for any errors after deployment.

Try these steps, and let me know if the problem persists!

**Comment by UtkarshK95** [comment]
The issue with `@apply` styles not working on Vercel can sometimes be attributed to the PostCSS processing setup or caching issues in Vercel. Here are some steps and configurations you can try to resolve this:

### 1. **Verify Tailwind and PostCSS Setup**

   - Ensure `tailwindcss` and `postcss` plugins are properly set up in `postcss.config.js`. Your config looks correct, but you can also try adding `postcss-nested` in case nested `@apply` rules are not being interpreted correctly.
   
   - Install `postcss-nested` by running:
     ```bash
     npm install -D postcss-nested
     ```
   
   - Update `postcss.config.js`:
     ```js
     export default {
       plugins: {
         'postcss-import': {},
         tailwindcss: {},
         autoprefixer: {},
         'postcss-nested': {}, // Add this line
       },
     };
     ```

### 2. **Check CSS Build Configuration**

   - Tailwind’s `@apply` directive relies on PostCSS to process the styles, and sometimes Vercel’s build might skip processing custom styles if caching is too aggressive. Make sure the `vite.config.js` is configured correctly:
     ```js
     import { defineConfig } from 'vite';
     import react from '@vitejs/plugin-react';
     import tailwindcss from 'tailwindcss';

     export default defineConfig({
       plugins: [react(), tailwindcss()],
     });
     ```

### 3. **Update Vercel’s Build Cache**

   - Vercel caches dependencies aggressively, which can sometimes cause issues. Add the `@apply`-related CSS file explicitly in your build script to ensure it gets processed each time. In your `package.json`, update your build command:
     ```json
     "build": "vite build --force"
     ```
   
   - `--force` will ensure that Vercel reprocesses all styles instead of using cached CSS files.

### 4. **Use CSS Preprocessing with Vite**

   - Vite has known limitations with handling nested `@apply` in certain scenarios. If the above steps don’t resolve the issue, try explicitly importing the `@apply` CSS file into a primary file, such as `index.css`, to ensure Vite processes it:
     ```css
     /* index.css */
     @import './assets/css/login.css';
     ```

### 5. **Test Locally with Production Build**

   - Before deploying, test locally by running:
     ```bash
     npm run build && npm run preview
     ```
   - This can reveal any configuration issues that might only occur in production builds.

After making these adjustments, deploy the project again to Vercel. This should help Vercel recognize and correctly apply Tailwind's `@apply` directives. If issues persist, clearing Vercel's cache and redeploying may also resolve the problem.
